### PR TITLE
Nishant's solutions uploaded

### PR DESCRIPTION
**Pull request by NishantPrabhu**
Solutions uploaded by Nishant Prabhu. Let me know about your suggestions.

### CODE REVIEW

**Comment by MukundVarmaT** [comment]
@NishantPrabhu Hey Thanks for putting up your solutions. 
You have provided a very detailed analysis of your solution. But I wanted to mention that one cannot benchmark the speed of a code by a single iteration of such a small number. (n = 5,7)

Also, I believe you have used the recursive method for calculating the numbers of the Fibonacci series which is comparatively much slower than a lot of other advanced techniques. One of the faster methods for this application is the Fast doubling method used very commonly in competitive programming. Please do check it out. 

I have taken the pleasure of benchmarking both of them for your reference for 1000 iterations and the corresponding time to calculate the ith Fibonacci number.

fibonacci-faster

Initially, till n = 40 the recursive method is slightly faster than the fast doubling method but for higher n the fast doubling method is exponentially faster.

Basically how the fast doubling method works is that:
Given F(k) and F(k+1), we can calculate these:

F(2k)     = F(k)[2F(k+1)−F(k)]
F(2k+1) = F(k+1)^2+F(k)^2

These identities can be extracted from the matrix exponentiation algorithm, which is another technique by which matrix multiplication is used to find the nth Fibonacci number.

Screenshot from 2020-03-25 18-42-20

The other method of either going line by line or calculating the entire list has its drawbacks as you have already mentioned on your file.

In case you want to check my code [link](https://github.com/MukundVarmaT/Daily-Sessions/blob/master/Day2/task-solutions/fibonacci-faster.py)

**Comment by MukundVarmaT** [comment]
Also regarding the second solution - you can still use the library in case you have repeating elements. You just need to ensure that there aren't any repeating elements which is done using the set command

```list(set(list(permutations([0, 0, 1, 2]))))```

**Comment by NishantPrabhu** [comment]
Thanks a lot for your suggestions. I'll keep these in mind.

Regards,
Nishant Prabhu

**Comment by jeeva2812** [comment]
@MukundVarmaT Fast Doubling method is faster to find the kth Fibonacci number but when finding all the Fibonacci numbers till k, won't its time complexity be O((n^2) log n), whereas the normal method will have a time complexity of only O(n^2)?

n - number of rows

**Comment by MukundVarmaT** [comment]
@jeeva2812 Hey you are right! My bad.

**Comment by MukundVarmaT** [comment]
@jeeva2812 Some memoization to make the whole calculation faster in case we want to find the nth Fibonacci number. This should be faster than the previous code I sent as well. Would love to hear your opinion

Code:
```
fibs = {0: 0, 1: 1}
def fib(n):
    if n in fibs: return fibs[n]
    if n % 2 == 0:
        fibs[n] = ((2 * fib((n / 2) - 1)) + fib(n / 2)) * fib(n / 2)
        return fibs[n]
    else:
        fibs[n] = (fib((n - 1) / 2) ** 2) + (fib((n+1) / 2) ** 2)
        return fibs[n]
```

This will be faster in case we need to find multiple Fibonacci numbers.

**Comment by jeeva2812** [comment]
Interesting solution!
